### PR TITLE
Feature/periodic build

### PR DESCRIPTION
**Pull request by pritamstyz4ever**
## Context

Currently when the node process terminates it happens abruptly and the in process messages are lost

## Objective

This PR adds a graceful shutdown plugin which can handle any task registered to be executed during termination

## References

https://github.com/screwdriver-cd/screwdriver/issues/1610

## License
I confirm that this contribution is made under a BSD license and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: lib/server.js [hunk 1]
```diff
@@ -144,8 +144,17 @@ module.exports = (config) => {
                     .generateProfile(username, scmContext, ['user'], metadata)
                 );
             server.app.jobFactory.executor.userTokenGen = server.app.jobFactory.tokenGen;
+            server.plugins.shutdown.handler({
+                taskname: 'executor-queue-cleanup',
+                task: async () => {
+                    await server.app.jobFactory.cleanUp();
+                    server.log(['shutdown'], 'completed clean up tasks', new Date().toISOString());
```
Comment (DekusDenial) [inline lib/server.js]: prefer to use `winston` for logging

File: lib/server.js [hunk 2]
```diff
@@ -144,8 +144,17 @@ module.exports = (config) => {
...
+                }
+            });
 
             // Start the server
-            return server.start().then(() => server);
+            return server.start()
+                .then(() => server)
+                .catch(err => console.log('Failed to start server', err));
```
Comment (DekusDenial) [inline lib/server.js]: same as above

File: lib/registerPlugins.js [hunk 3]
```diff
@@ -32,6 +32,20 @@ function registerDefaultPlugins(server, callback) {
     }, callback);
 }
 
+/**
+ *
```
Comment (DekusDenial) [inline lib/registerPlugins.js]: consider adding little bit more stuff for jsdoc

File: plugins/shutdown.js [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    return '';
+}
+
+/**
+ *
+ * @param {function} fn
+ * @param {string} timeout
+ */
+function promiseTimeout(fn, timeout) {
+    return new Promise(((resolve, reject) => {
```
Comment (DekusDenial) [inline plugins/shutdown.js]: not sure which is more elegant, this or using `Promise.race`?

File: plugins/shutdown.js [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+ * Hapi interface for plugin to handle serve graceful shutdown
+ * @method register
+ * @param  {Hapi.Server}    server
+ * @param  {Object}         options
+ */
+exports.register = (server, options, next) => {
+    const taskHandler = async (resolve, reject) => {
+        try {
+            await Promise.all(Object.keys(tasks).map(async (key) => {
+                server.log(['shutdown'], `executing task ${key}`, new Date().toISOString());
```
Comment (DekusDenial) [inline plugins/shutdown.js]: same as above

File: plugins/shutdown.js [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+            resolve();
+        } catch (err) {
+            console.log(err);
+            reject(err);
+        }
+    };
+    const gracefulStop = async () => {
+        try {
+            server.log(['shutdown'], 'gracefully shutting down server', new Date().toISOString());
+            await server.root.stop({
```
Comment (DekusDenial) [inline plugins/shutdown.js]: what's the difference btw this and `server.stop()`?

File: plugins/shutdown.js [hunk 7]
```diff
@@ -24,19 +33,21 @@ function register(task) {
...
-        // Set up the timeout
-        setTimeout(() => {
-            resolve(`Promise timed out after ${timeout} ms`);
-        }, timeout);
-    }));
+    return Promise.race([
+        Promise.resolve(fn),
+        new Promise((resolve) => {
+            const id = setTimeout(() => {
+                clearTimeout(id);
```
Comment (DekusDenial) [inline plugins/shutdown.js]: It seems unnecessary to call clear timeout

Comment (pritamstyz4ever) [reply plugins/shutdown.js]: Yeah, must be a left over

**Pull request merged**